crowbar_framework/lib/upgrade.rb

```diff
@@ -0,0 +1,64 @@
+class Upgrade
+  attr_accessor :current_step, :substep, :current_node, :upgrade_progress
+  attr_reader :steps
+
+  # Return the current state of upgrade process.
+  # We're keeping the information in the file so is accessible by other applications
+  # and different crowbar versions.
+  def initialize
+    @upgrade_progress = {
+      current_step: upgrade_steps_6_7.first,
+      # substep is needed for more complex steps like upgrading the nodes
+      substep: "",
+      # current node is relevant only for the nodes_upgrade step
+      current_node: ""
+    }
+    if progress_file_path.exist?
+      @upgrade_progress = JSON.load(progress_file_path.read)
+    else
+      # in 'steps', we save the information about each step that was executed
+      @upgrade_progress["steps"] = upgrade_steps_6_7.map do |step|
+        [step, { status: "pending", errors: {} }]
+      end.to_h
+    end
+    @current_step = @upgrade_progress["current_step"]
```

Comment:
But I could use separate methods to manipulate with steps, substeps etc. just like @rsalevsky proposed. I do not mind which solution we're gonna use, just please agree on one.